### DISCUSSION TITLE
Sprite colors wrong, TFT colors OK on Arduino Due, ILI9488 SPI

### DISCUSSION THREAD

**Discussion by PatMan6889** [Bodmer/TFT_eSPI#3595] [Q&A - Sprites] [answered]
Hi,
using 
`tft.fillScreen(TFT_BLUE);`
or any other TFT_eSPI functions delivers correct colors.

When I like to add a sprites:
`sprite.fillCircle(100, 100, 20, TFT_YELLOW);`
the colors are wrong:
Blue is green. Green is red. Red is Blue. Yellow is magenta.
The resulting colors are a bit off and don't exactly match the colors on tft.
Working colors are only black and white.

I tried to play around with setSwapBytes and setColorDepth.
setSwapBytes changes nothing. setColorDepth changes the colors but not to the correct state.

I use the same code as in examples but get different results.
Any ideas?

**Comment by Basitadam** [comment]
Hi,
The order of setSwapBytes is important. You need to do it before creating the sprite. Like this
```
  Needle.setSwapBytes(true);
  Needle.setColorDepth(16);
  Needle.createSprite(135,68);                      // create the needle Sprite
  Needle.fillSprite(TFT_BLACK);  
```

**Reply by PatMan6889** [reply]
Unfortunately when I try 
```
  txtSprite.setSwapBytes(true);
  txtSprite.setColorDepth(16);
  txtSprite.createSprite(100, 30);
  txtSprite.fillSprite(TFT_YELLOW);
  txtSprite.setTextColor(TFT_WHITE, TFT_BLACK);
  txtSprite.drawString("123", 0, 0);
  txtSprite.pushSprite(0, 0);
  txtSprite.deleteSprite();
```
I get a kind of magenta filled sprite with the string 123 printed in white with a black text background.
I would expect a yellow filled sprite instead of magenta there. Black and white are working fine, all other colors don't match.

Maybe it is a hardware limitation of the Arduino Due?

**Reply by Basitadam** [reply]
with 
```
txtSprite.setColorDepth(8);
```
instead?

**Reply by PatMan6889** [reply]
> with
> 
> ```
> txtSprite.setColorDepth(8);
> ```
> 
> instead?

I get exactly the same result as with setColorDepth(16)

**Comment by Duiker-cpu** [comment]
Hier ondervind ik exact dezelfde problemen. Ook met een ILI9488 SPI display.
Ik probeerde het volgende: Op de linkerhelft 3 kleurenbalken onder elkaar tekenen. Rood, Groen en Blauw.
Dan een blok per kleur naar een sprite copieren, en de sprite rechts plaatsen. Dat werkt per kleur perfect.
Maar het gaat mis als ik een geladen plaatje plaats. Dat plaatje wordt perfect weergegeven. Maar de sprite's hebben dan de verkeerde kleuren. Met ChatGPT vanalles geprobeerd. Het lijkt aan het plaatje te liggen.

**Comment by PatMan6889** [comment, marked as answer]
I found the problem. The colors need to be given in BRG instead of RGB.
Maybe there is a better solution but this function converts the color:

`
/**
 * @brief Converts a 16-bit RGB color value to BRG (Blue-Red-Green).
 * 
 * This function takes a 16-bit color value in RGB 565 format (5 bits Red, 6 bits Green, 5 bits Blue) 
 * and converts it to BRG format by rearranging the color components.
 *
 * @param color A 16-bit color value in RGB 565 format.
 * @return uint16_t The converted color value in BRG 565 format.
 */
uint16_t rgb_to_brg(uint16_t color) {
    // Extract individual color components (5-6-5 bits)
    uint8_t r5 = (color >> 11) & 0x1F;  // 5 bits for Red
    uint8_t g6 = (color >> 5) & 0x3F;   // 6 bits for Green
    uint8_t b5 = color & 0x1F;          // 5 bits for Blue

    // Scaling to 8 bits (0-255)
    r5 = (r5 * 527 + 23) >> 6;  // 527 = (255 + 1) * 2^6 / (2^5 - 1)
    g6 = (g6 * 259 + 33) >> 6;  // 259 = (255 + 1) * 2^6 / (2^6 - 1)
    b5 = (b5 * 527 + 23) >> 6;  // 527 = (255 + 1) * 2^6 / (2^5 - 1)
    
    // Convert back to 5-6-5 bits and rearrange color components to BRG
    return ((b5 & 0xF8) << 8) | ((r5 & 0xFC) << 3) | (g6 >> 3);;
}
`

**Reply by Basitadam** [reply]
Hi. Can you test this:

Before sprite commands run this:
```
   tft.writecommand(0x0b); // Memory Access Control
   tft.writedata(0x4c);          // MX, BGR
```
After sprite commands run this:

```
    tft.writecommand(0x0b); // Memory Access Control
    tft.writedata(0x48);          // MX, RGB
```

or the other way around.

**Reply by PatMan6889** [reply]
> Hi. Can you test this:
> 
> Before sprite commands run this:
> 
> ```
>    tft.writecommand(0x0b); // Memory Access Control
>    tft.writedata(0x4c);          // MX, BGR
> ```
> 
> After sprite commands run this:
> 
> ```
>     tft.writecommand(0x0b); // Memory Access Control
>     tft.writedata(0x48);          // MX, RGB
> ```
> 
> or the other way around.

I tried it both ways. There is no change. Colors stil wrong.

**Reply by Basitadam** [reply]
Sorry. Like this:
```
   tft.writecommand(0x0b); // Memory Access Control
   tft.writedata(0x40);          // MX, RGB
 ```

and

```
   tft.writecommand(0x0b); // Memory Access Control
   tft.writedata(0x48);          // MX, BGR
 ```

**Reply by PatMan6889** [reply]
Still no success

**Reply by Basitadam** [reply]
Last try:
```
   tft.writecommand(0x36); // Memory Access Control
   tft.writedata(0x40);          // MX, RGB
```

and

```
   tft.writecommand(0x36); // Memory Access Control
   tft.writedata(0x48);          // MX, BGR
```

**Reply by PatMan6889** [reply]
Still not working. The display is kind of rotated.
One thing I see in the comment is // MX, BGR. I need to convert to BRG.

**Reply by Basitadam** [reply]
This is the register. We are changing only the color selector switch control bit. Should have worked:

![image](https://github.com/user-attachments/assets/bba2d72b-21fb-4572-a71d-dbdce7a4d4b3)
